PageRank is incorrect

PageRank does not take dangling nodes into account.

It seems the PageRank implementation is not complete as it does not take dangling vertices into account (https://github.com/GraphBLAS/LAGraph/issues/9):
![image](https://user-images.githubusercontent.com/1402801/63653803-f2c8f980-c771-11e9-8284-1aa46254193d.png)

A correct solution in terms of linear algebra (implemented in Julia) is as follows:

```julia
out = A*ones(N)
D = map(x -> x == 0 ? 1.0 : 0.0, out)
pr = ones(N)/N
for i in 1:iterations
    teleport = (1-d) / N
    importance = d * A'*(pr./out)
    dangling = d/N * sum(pr.*D)
    global pr = teleport .+ importance .+ dangling
end
```

The current PageRank results are definitely incorrect as they have negative values.

```console
java.lang.AssertionError: vertex 1 has correct value
Expected: is a numeric value within <1.0E-4> of <0.0>
     but: <-0.7257517637024848> differed by <0.7256517637024849>

	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
	at science.atlarge.graphalytics.validation.algorithms.pr.PageRankValidationTest.validatePageRank(PageRankValidationTest.java:115)
	at science.atlarge.graphalytics.validation.algorithms.pr.PageRankValidationTest.testDirectedPageRankOnValidationGraph(PageRankValidationTest.java:82)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:47)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:44)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:70)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:50)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:160)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
```

--- a/pom.xml
+++ b/pom.xml
@@ -8,7 +8,7 @@
 
     <properties>
         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
-        <graphalytics.version>1.1.0</graphalytics.version>
+        <graphalytics.version>1.2.0-SNAPSHOT</graphalytics.version>
         <flink.version>1.9.0</flink.version>
         <hadoop.version>2.4.1</hadoop.version>
         <log4j.version>2.5</log4j.version>

--- a/pom.xml
+++ b/pom.xml
@@ -118,8 +118,8 @@
                 <artifactId>maven-compiler-plugin</artifactId>
                 <version>3.1</version>
                 <configuration>
-                    <source>1.7</source>
-                    <target>1.7</target>
+                    <source>1.8</source>
+                    <target>1.8</target>
                 </configuration>
             </plugin>
 

--- a/src/main/java/science/atlarge/graphalytics/flink/GellyPlatform.java
+++ b/src/main/java/science/atlarge/graphalytics/flink/GellyPlatform.java
@@ -188,8 +188,7 @@ public class GellyPlatform extends AbstractPlatform {
 					new LocalClusteringCoefficient(isDirected), hasEdgeValues);
 				break;
 			case "PR": job = new GellyJob<>(remoteEnv, vertexPath, edgesPath, outputPath,
-					new ScatterGatherPageRank(
-							parameters, input.getNumberOfVertices()), hasEdgeValues);
+					new ScatterGatherPageRank(parameters), hasEdgeValues);
 				break;
 			case "SSSP": job = new GellyJob<>(remoteEnv, vertexPath, edgesPath, outputPath,
 					new ScatterGatherSSSP(parameters), hasEdgeValues);

--- a/src/main/java/science/atlarge/graphalytics/flink/algorithms/pr/ScatterGatherPageRank.java
+++ b/src/main/java/science/atlarge/graphalytics/flink/algorithms/pr/ScatterGatherPageRank.java
@@ -18,6 +18,7 @@
 
 package science.atlarge.graphalytics.flink.algorithms.pr;
 
+import org.apache.flink.api.common.aggregators.DoubleSumAggregator;
 import org.apache.flink.api.common.functions.MapFunction;
 import org.apache.flink.api.java.DataSet;
 import org.apache.flink.api.java.tuple.Tuple2;
@@ -26,29 +27,35 @@ import org.apache.flink.graph.EdgeJoinFunction;
 import org.apache.flink.graph.Graph;
 import org.apache.flink.graph.GraphAlgorithm;
 import org.apache.flink.graph.Vertex;
+import org.apache.flink.graph.asm.degree.annotate.directed.VertexOutDegree;
 import org.apache.flink.graph.spargel.GatherFunction;
 import org.apache.flink.graph.spargel.MessageIterator;
 import org.apache.flink.graph.spargel.ScatterFunction;
+import org.apache.flink.graph.spargel.ScatterGatherConfiguration;
+import org.apache.flink.types.DoubleValue;
 import org.apache.flink.types.LongValue;
 import org.apache.flink.types.NullValue;
 import science.atlarge.graphalytics.domain.algorithms.AlgorithmParameters;
 import science.atlarge.graphalytics.domain.algorithms.PageRankParameters;
 
+
 public class ScatterGatherPageRank implements GraphAlgorithm<Long, NullValue, NullValue, DataSet<Tuple2<Long, Double>>> {
 
-	private final float beta;
+	private final float damping;
 	private final int maxIterations;
-	private final long numberOfVertices;
 
-	public ScatterGatherPageRank(AlgorithmParameters params, long numVertices) {
+	public ScatterGatherPageRank(AlgorithmParameters params) {
 		PageRankParameters prParams = (PageRankParameters)params;
-		beta = prParams.getDampingFactor();
+		damping = prParams.getDampingFactor();
 		maxIterations = prParams.getNumberOfIterations();
-		numberOfVertices = numVertices;
 	}
 
 	@Override
-	public DataSet<Tuple2<Long, Double>> run(Graph<Long, NullValue, NullValue> network) {
+	public DataSet<Tuple2<Long, Double>> run(Graph<Long, NullValue, NullValue> network) throws Exception {
+		DataSet<Vertex<Long, LongValue>> outDegrees = network.run(new VertexOutDegree<>());
+
+		// determine the number of dangling vertices: vertexOutDegrees only contains non-dangling vertices
+		long numberOfDanglingVertices = network.numberOfVertices() - outDegrees.count();
 
 		DataSet<Tuple2<Long, LongValue>> vertexOutDegrees = network.outDegrees();
 
@@ -57,10 +64,16 @@ public class ScatterGatherPageRank implements GraphAlgorithm<Long, NullValue, Nu
 				.mapEdges(new InitEdgeValues())
 				.joinWithEdgesOnSource(vertexOutDegrees, new InitWeights());
 
+		ScatterGatherConfiguration parameters = new ScatterGatherConfiguration();
+		parameters.setOptDegrees(true);
+		parameters.setOptNumVertices(true);
+		parameters.registerAggregator("danglingAggregator", new DoubleSumAggregator());
+
 		return networkWithWeights.runScatterGatherIteration(
-				new RankMessenger(numberOfVertices),
-				new VertexRankUpdater(beta, numberOfVertices),
-				maxIterations)
+				new RankMessenger(),
+				new VertexRankUpdater(damping, numberOfDanglingVertices),
+				maxIterations,
+				parameters)
 			.getVertices();
 	}
 
@@ -71,24 +84,42 @@ public class ScatterGatherPageRank implements GraphAlgorithm<Long, NullValue, Nu
 	@SuppressWarnings("serial")
 	public static final class VertexRankUpdater<K> extends GatherFunction<K, Double, Double> {
 
-		private final float beta;
-		private final long numVertices;
-		
-		public VertexRankUpdater(float beta, long numberOfVertices) {
-			this.beta = beta;
-			this.numVertices = numberOfVertices;
+		private final float damping;
+		private final double numberOfDanglingVertices;
+
+		public VertexRankUpdater(float damping, double numberOfDanglingVertices) {
+			this.damping = damping;
+			this.numberOfDanglingVertices = numberOfDanglingVertices;
 		}
 
 		@Override
 		public void updateVertex(Vertex<K, Double> vertex, MessageIterator<Double> inMessages) {
 			double rankSum = 0.0;
 			for (double msg : inMessages) {
-				rankSum = msg;
+				rankSum += msg;
 			}
 
-			// apply the dampening factor / random jump
-			double newRank = (beta * rankSum) + (1.0 - beta) / (double) numVertices;
+			double sumDangling;
+			if (getSuperstepNumber() == 1) {
+				sumDangling = numberOfDanglingVertices / (double) getNumberOfVertices();
+			} else {
+				DoubleValue sumDanglingValue = getPreviousIterationAggregate("danglingAggregator");
+				sumDangling = sumDanglingValue.getValue();
+			}
+
+			// compute new PageRank value: teleport + importance + dangling
+			double teleport = (1.0 - damping) / (double) getNumberOfVertices();
+			double importance = damping * rankSum;
+			double dangling = damping / (double) getNumberOfVertices() * sumDangling;
+
+			double newRank = teleport + importance + dangling;
 			setNewVertexValue(newRank);
+
+			// for dangling vertices, aggregate their PageRank
+			if (getOutDegree() == 0) {
+				DoubleSumAggregator danglingAggregator = getIterationAggregator("danglingAggregator");
+				danglingAggregator.aggregate(newRank);
+			}
 		}
 	}
 
@@ -100,22 +131,20 @@ public class ScatterGatherPageRank implements GraphAlgorithm<Long, NullValue, Nu
 	@SuppressWarnings("serial")
 	public static final class RankMessenger<K> extends ScatterFunction<K, Double, Double, Double> {
 
-		private final long numVertices;
-
-		public RankMessenger(long numberOfVertices) {
-			this.numVertices = numberOfVertices;
+		public RankMessenger() {
 		}
 
 		@Override
 		public void sendMessages(Vertex<K, Double> vertex) {
 			if (getSuperstepNumber() == 1) {
 				// initialize vertex ranks
-				vertex.setValue(new Double(1.0 / (double) numVertices));
+				vertex.setValue(1.0 / (double) getNumberOfVertices());
 			}
 
 			for (Edge<K, Double> edge : getEdges()) {
 				sendMessageTo(edge.getTarget(), vertex.getValue() * edge.getValue());
 			}
+			sendMessageTo(vertex.getId(), 0.0);
 		}
 	}
 

--- a/src/test/java/gelly/graphalytics/GatherScatterPageRankJobTest.java
+++ b/src/test/java/gelly/graphalytics/GatherScatterPageRankJobTest.java
@@ -1,21 +1,23 @@
 package gelly.graphalytics;
 
-import org.junit.Ignore;
-import science.atlarge.graphalytics.domain.algorithms.PageRankParameters;
-import science.atlarge.graphalytics.flink.algorithms.pr.ScatterGatherPageRank;
-import science.atlarge.graphalytics.validation.GraphStructure;
-import science.atlarge.graphalytics.validation.algorithms.pr.PageRankOutput;
-import science.atlarge.graphalytics.validation.algorithms.pr.PageRankValidationTest;
 import org.apache.flink.api.java.DataSet;
 import org.apache.flink.api.java.ExecutionEnvironment;
 import org.apache.flink.api.java.tuple.Tuple2;
 import org.apache.flink.graph.Edge;
 import org.apache.flink.graph.Graph;
 import org.apache.flink.types.NullValue;
+import science.atlarge.graphalytics.domain.algorithms.PageRankParameters;
+import science.atlarge.graphalytics.flink.algorithms.pr.ScatterGatherPageRank;
+import science.atlarge.graphalytics.validation.GraphStructure;
+import science.atlarge.graphalytics.validation.algorithms.pr.PageRankOutput;
+import science.atlarge.graphalytics.validation.algorithms.pr.PageRankValidationTest;
 
-import java.util.*;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
 
-@Ignore
 public class GatherScatterPageRankJobTest extends PageRankValidationTest {
 
     @Override
@@ -24,8 +26,7 @@ public class GatherScatterPageRankJobTest extends PageRankValidationTest {
 
         Graph<Long, NullValue, NullValue> input = getInputGraph(graphStructure);
         // run the PageRank job
-        DataSet<Tuple2<Long, Double>> result = input.run(
-                new ScatterGatherPageRank(params, graphStructure.getVertices().size()));
+        DataSet<Tuple2<Long, Double>> result = input.run(new ScatterGatherPageRank(params));
         return convertResult(result);
     }
 
@@ -35,8 +36,7 @@ public class GatherScatterPageRankJobTest extends PageRankValidationTest {
 
         Graph<Long, NullValue, NullValue> input = getInputGraph(graphStructure);
         // run the PageRank job
-        DataSet<Tuple2<Long, Double>> result = input.run(
-                new ScatterGatherPageRank(params, graphStructure.getVertices().size()));
+        DataSet<Tuple2<Long, Double>> result = input.run(new ScatterGatherPageRank(params));
         return convertResult(result);
     }
 

